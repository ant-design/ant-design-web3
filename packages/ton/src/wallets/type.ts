import type { Wallet, WalletMetadata } from '@ant-design/web3-common';
import type { WalletInfo } from '@tonconnect/sdk';

export type TonBasicWallet = TonWalletMetadata &
  WalletInfo & {
    universalLink?: string;
  };

export type TonWallet = Wallet & WalletInfo;

export interface WalletFactory {
  create: () => TonWallet;
}

export type TonWalletMetadata = Omit<WalletMetadata, 'name' | 'remark'>;
