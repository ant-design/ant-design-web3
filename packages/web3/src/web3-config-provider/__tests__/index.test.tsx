import React, { useContext } from 'react';
import {
  ConfigContext,
  ConnectButton,
  useAccount,
  useNFT,
  Web3ConfigProvider,
  type Account,
} from '@ant-design/web3';
import type { Web3ConfigProviderProps } from '@ant-design/web3-common';
import { fireEvent, render } from '@testing-library/react';
import { Spin } from 'antd';
import { describe, expect, it, vi } from 'vitest';

const baseProps: Web3ConfigProviderProps = {
  availableWallets: [
    {
      name: 'MetaMask',
      remark: 'Easy-to-use browser extension.',
      extensions: [
        {
          key: 'Chrome',
          browserIcon:
            'https://github.com/ant-design/ant-design/assets/10286961/0d4e4ac7-8f89-4147-a06a-de72c02e85cb',
          browserName: 'Chrome',
          link: 'https://chromewebstore.google.com/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn',
          description: 'Access your wallet right from your favorite web browser.',
        },
      ],
    },
  ],
};

describe('web3-config-provider', () => {
  it('Should work', () => {
    const App: React.FC = () => {
      const [account, setAccount] = React.useState<Account>();
      return (
        <Web3ConfigProvider
          {...baseProps}
          connect={async () => {
            const newAccount = {
              address: '0x1234567890123456789012345678901234567890',
            };
            setAccount(newAccount);
          }}
          account={account}
        >
          <ConnectButton />
        </Web3ConfigProvider>
      );
    };

    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });
  it('useAccount', () => {
    const Child: React.FC = () => {
      const { account } = useAccount();
      return <div>{account?.address}</div>;
    };
    const App: React.FC = () => {
      const [account, setAccount] = React.useState<Account>();
      return (
        <Web3ConfigProvider
          {...baseProps}
          connect={async () => {
            const newAccount = {
              address: '0x1234567890123456789012345678901234567890',
            };
            setAccount(newAccount);
          }}
          account={account}
        >
          <Child />
        </Web3ConfigProvider>
      );
    };

    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });
  it('useNFT', () => {
    const Child: React.FC = () => {
      const { metadata, error, loading } = useNFT(
        '0x79fcdef22feed20eddacbb2587640e45491b757f',
        42n,
      );
      if (error) {
        return <div>{error.message}</div>;
      }
      return <Spin spinning={loading}>{metadata.name}</Spin>;
    };
    const App: React.FC = () => {
      const [account, setAccount] = React.useState<Account>();
      return (
        <Web3ConfigProvider
          {...baseProps}
          connect={async () => {
            const newAccount = {
              address: '0x1234567890123456789012345678901234567890' as `0x${string}`,
            };
            setAccount(newAccount);
          }}
          account={account}
        >
          <Child />
        </Web3ConfigProvider>
      );
    };

    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });
  it('useNFT with custom getNFTMetadata', () => {
    const Child: React.FC = () => {
      const { metadata, error, loading } = useNFT(
        '0x79fcdef22feed20eddacbb2587640e45491b757f',
        42n,
      );
      if (error) {
        return <div>{error.message}</div>;
      }
      return <Spin spinning={loading}>{metadata.name}</Spin>;
    };
    const App: React.FC = () => {
      const [account, setAccount] = React.useState<Account>();
      return (
        <Web3ConfigProvider
          {...baseProps}
          connect={async () => {
            const newAccount = {
              address: '0x1234567890123456789012345678901234567890' as `0x${string}`,
            };
            setAccount(newAccount);
          }}
          account={account}
          getNFTMetadata={() => {
            return Promise.resolve({
              attributes: [],
              name: 'test',
              date: Date.now(),
              description: 'test',
              image: 'https://no.com/z.png',
            });
          }}
        >
          <Child />
        </Web3ConfigProvider>
      );
    };

    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });
  it('extendable', () => {
    const Child: React.FC = () => {
      const { account, balance } = useContext(ConfigContext);
      return (
        <>
          <div id="account-name">{account?.name}</div>
          <div id="balance">{balance?.value?.toString()}</div>
        </>
      );
    };

    const App: React.FC = () => {
      return (
        <Web3ConfigProvider
          account={{
            name: 'testAccount',
            address: '0x123456789',
          }}
          balance={{ value: 200n }}
        >
          <Web3ConfigProvider balance={{ value: 100n }}>
            <Child />
          </Web3ConfigProvider>
        </Web3ConfigProvider>
      );
    };

    const { baseElement } = render(<App />);
    expect(baseElement.querySelector('#account-name')?.textContent).toBe('testAccount');
    expect(baseElement.querySelector('#balance')?.textContent).toBe('100');
  });
  it('should not extend from parent', () => {
    const Child: React.FC = () => {
      const { account, balance } = useContext(ConfigContext);
      return (
        <>
          <div id="account-name">{account?.name}</div>
          <div id="balance">{balance?.value?.toString()}</div>
        </>
      );
    };

    const App: React.FC = () => {
      return (
        <Web3ConfigProvider
          account={{
            name: 'testAccount',
            address: '0x123456789',
          }}
          balance={{ value: 200n }}
        >
          <Web3ConfigProvider extendsContextFromParent={false} balance={{ value: 100n }}>
            <Child />
          </Web3ConfigProvider>
        </Web3ConfigProvider>
      );
    };

    const { baseElement } = render(<App />);
    expect(baseElement.querySelector('#account-name')?.textContent).toBeFalsy();
    expect(baseElement.querySelector('#balance')?.textContent).toBe('100');
  });
  it('merge locale', async () => {
    const { baseElement } = render(
      <Web3ConfigProvider
        locale={{
          ConnectButton: {
            disconnect: 'Parent Disconnect',
            copyAddress: 'Parent Copy Address',
          },
        }}
      >
        <Web3ConfigProvider
          locale={{
            ConnectButton: {
              copyAddress: 'Child Copy Address',
            },
          }}
        >
          <ConnectButton
            account={{
              address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
            }}
          />
        </Web3ConfigProvider>
      </Web3ConfigProvider>,
    );

    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button') as Element);
    await vi.waitFor(() => {
      expect(
        baseElement.querySelector('.ant-modal-footer')!.querySelectorAll('.ant-btn')[0]
          .childNodes[0].textContent,
      ).toBe('Child Copy Address');
      expect(
        baseElement.querySelector('.ant-modal-footer')!.querySelectorAll('.ant-btn')[1]
          .childNodes[0].textContent,
      ).toBe('Parent Disconnect');
    });
  });
});
