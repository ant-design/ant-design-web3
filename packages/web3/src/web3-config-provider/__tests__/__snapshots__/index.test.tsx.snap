// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`web3-config-provider > Should work 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1c0na6j ant-btn-default ant-web3-connect-button css-dev-only-do-not-override-1c0na6j"
      type="button"
    >
      <div
        class="ant-web3-connect-button-content"
      >
        <div
          class="ant-web3-connect-button-content-inner"
        >
          <div
            class="ant-web3-connect-button-text"
          >
            Connect Wallet
          </div>
        </div>
      </div>
    </button>
  </div>
</body>
`;

exports[`web3-config-provider > useAccount 1`] = `
<body>
  <div>
    <div />
  </div>
</body>
`;

exports[`web3-config-provider > useNFT 1`] = `
<body>
  <div>
    <div>
      Provider is not ready
    </div>
  </div>
</body>
`;

exports[`web3-config-provider > useNFT with custom getNFTMetadata 1`] = `
<body>
  <div>
    <div
      aria-busy="false"
      aria-live="polite"
      class="ant-spin css-dev-only-do-not-override-1c0na6j"
    >
      <span
        class="ant-spin-dot-holder"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </span>
    </div>
  </div>
</body>
`;
